### PR TITLE
Elasticsearch Getting Started Guide

### PR DESCRIPTION
**Pull request by ackerman80**
Signed-off-by: Arvind Soni <arvindsoni@gmail.com>

### CODE REVIEW

**Comment by ackerman80** [comment]
test-docs-please

**Comment by ackerman80** [comment]
test-me-please

**Comment by ackerman80** [comment]
Tested on K8s 1.8, 1.9, 1.10. Partial test on 1.7 that the app yaml works correctly

File: Documentation/gettingstarted/elasticsearch.rst [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+.. include:: minikube_intro.rst
+.. include:: cilium_install.rst
+
+Step 2: Deploy the Demo Application
+===================================
+
+Now that we have Cilium running, we can deploy our demo Elasticsearch application. The demo application aligns with our tradition of using Star Wars-themed examples. We go back to the time when Vader had recently been converted to the Dark side. Darth Sidious wanted to share some of the books that he had authored. The books were stored in an Elasticsearch database and exposed via a service of the same name. Both Sidious and Vader accessed these books using python-based clients running in Kubernetes pods.
+
+.. image:: images/cilium_es_gsg_topology.png
+   :scale: 30 %
```
Comment (aanm) [inline Documentation/gettingstarted/elasticsearch.rst]: Is 40% to big?

File: Documentation/gettingstarted/elasticsearch.rst [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+The file ``es-sw-app.yaml`` will deploy the Elasticsearch service which stores Sidious' books and it will create one Vader and one Sidious client pod each.
+
+.. parsed-literal::
+
+    $ kubectl create -f \ |SCM_WEB|\/examples/kubernetes-es/es-sw-app.yaml
+    serviceaccount "elasticsearch" created
+    service "elasticsearch" created
+    replicationcontroller "es" created
+    role "elasticsearch" created
```
Comment (aanm) [inline Documentation/gettingstarted/elasticsearch.rst:30]: The output  was different for me, specially the "deployment"
```
    serviceaccount "elasticsearch" created
    service "elasticsearch" created
    replicationcontroller "es" created
    role.rbac.authorization.k8s.io "elasticsearch" created
    rolebinding.rbac.authorization.k8s.io "elasticsearch" created
    pod "sidious" created
    pod "vader" created
```

File: Documentation/gettingstarted/elasticsearch.rst [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    po/es-bwvnp        1/1       Running   0          1m
+    po/etcd-cilium-0   1/1       Running   0          7m
+    po/sidious         1/1       Running   0          1m
+    po/vader           1/1       Running   0          1m
+
+
+Step 3: Security Risks for Elasticsearch Access
+===============================================
+
+A fundamental security concern for Elasticsearch service is *which client services should be allowed to access what content and perform what actions*. This is an access control problem at the API-layer (i.e L7-layer). In this example, the security challenge for Darth Sidious is that he does not trust Vader, a newly converted apprentice. So he is very worried that Vader can manipulate and ``PUT`` new versions of his books! Sidious wants Vader to have only ``GET`` access including the ability to *search* the database. But he does not want Vader to have ``PUT`` access. Run the following commands to see that both Sidious and Vader have ``GET`` and ``PUT`` access to the Elasticsearch service.
```
Comment (aanm) [inline Documentation/gettingstarted/elasticsearch.rst]: `which client....` is in italic, you meant italic or bold?

File: Documentation/gettingstarted/elasticsearch.rst [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+::
+
+    $ kubectl exec sidious -- python create.py
+    Creating/Updating Books
+    created :  {'_index': 'sidious', '_type': 'tome', '_id': '1', '_version': 1, 'result': 'created', '_shards': {'total': 2, 'successful': 1, 'failed': 0}, 'created': True}
```
Comment (aanm) [inline Documentation/gettingstarted/elasticsearch.rst:59]: Output was different, I got 2 created messages
```
Creating/Updating Books
created :  {'_index': 'sidious', '_type': 'tome', '_id': '1', '_version': 1, 'result': 'created', '_shards': {'total': 2, 'successful': 1, 'failed': 0}, 'created': True}
created :  {'_index': 'sidious', '_type': 'tome', '_id': '2', '_version': 1, 'result': 'created', '_shards': {'total': 2, 'successful': 1, 'failed': 0}, 'created': True}
```

File: Documentation/gettingstarted/elasticsearch.rst [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+::
+
+    $ kubectl exec vader -- python update.py
+    Creating/Updating Books
+    updated :  {'_index': 'sidious', '_type': 'tome', '_id': '1', '_version': 3, 'result': 'updated', '_shards': {'total': 2, 'successful': 1, 'failed': 0}, 'created': False}
+    updated :  {'_index': 'sidious', '_type': 'tome', '_id': '2', '_version': 3, 'result': 'updated', '_shards': {'total': 2, 'successful': 1, 'failed': 0}, 'created': False}
+
+    $ kubectl exec vader -- python get_search.py
+    Searching for Books by Darth Sidious
+    Got 2 Hits:
```
Comment (aanm) [inline Documentation/gettingstarted/elasticsearch.rst:82]: Output is different:
```
Searching for Books by Darth Sidious
Got 2 Hits:
{'_index': 'sidious', '_type': 'tome', '_id': '2', '_score': 1.0, '_source': {'author': 'sidious', 'title': 'Force is Same for Dark Side and Jedi'}}
{'_index': 'sidious', '_type': 'tome', '_id': '1', '_score': 1.0, '_source': {'author': 'sidious', 'title': 'Why Convert a Jedi!'}}
Get Book 1 by Darth Sidious
{'author': 'sidious', 'title': 'Why Convert a Jedi!'}
```

File: Documentation/gettingstarted/elasticsearch.rst [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    $ kubectl exec sidious -- python get_search.py
+    Searching for Books by Darth Sidious
+    Got 2 Hits:
+    {'_index': 'sidious', '_type': 'tome', '_id': '2', '_score': 1.0, '_source': {'author': 'sidious', 'title': 'Welcome to the Dark Side'}}
+    {'_index': 'sidious', '_type': 'tome', '_id': '1', '_score': 1.0, '_source': {'author': 'sidious', 'title': 'Convert Jedi to Dark Side: 101'}}
+    Get Book 1 by Darth Sidious
+    {'author': 'sidious', 'title': 'Convert Jedi to Dark Side: 101'}
+
+
+Sidious has access to perform ``GET`` and ``PUT`` as expected. But see what happens when Vader has both ``GET`` and ``PUT`` access. Vader can completely modify the books! (Note the change in book titles e.g. "Why convert a Jedi!")
```
Comment (aanm) [inline Documentation/gettingstarted/elasticsearch.rst]: I would change it to `Vader can, and did completely modified the books! (Note the change in book titles from 'Convert Jedi to Dark Side: 101' to "Why convert a Jedi!")`

File: examples/kubernetes-es/es-sw-policy.yaml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    - ports:
+      - port: "9200"
+        protocol: TCP
+      rules:
+        http:
+        - method: ^GET$
+          path: ^/sidious/tome/[^/]+/??[^/]*$
+        - method: ^GET$
+          path: ^/sidious/_search/??[^/]*$
+- egress:
```
Comment (aanm) [inline examples/kubernetes-es/es-sw-policy.yaml:42]: ~Do we need the egress policy for this example?~

Yes we do, for kube-dns

**Review by aanm — CHANGES_REQUESTED** [review]

File: Documentation/gettingstarted/elasticsearch.rst [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+===========================================
+
+Fortunately for Darth Sidious, the Empire DevOps team is using Cilium for their Kubernetes cluster. Cilium provides L7 visibility and security policies to control Elasticsearch API access. In this case, Sidious orders to get the following policy pushed which gives him both ``GET`` and ``PUT`` access to his pods but restricts Vader's pods to only ``GET`` access.
+
+.. literalinclude:: ../../examples/kubernetes-es/es-sw-policy.yaml
+
+
+Cilium follows the white-list, least privilege model for security. A *CiliumNetworkPolicy* contains a list of rules that define allowed requests, meaning that requests that do not match any rules are denied.
+
+In this example, the policy rules are defined for inbound traffic (i.e., "ingress") connections to the *elasticsearch* service. Note that endpoints selected as backend pods for the service are defined by the *selector* labels. *Selector* labels use the same concept as Kubernetes to define a service. In this example, label ``k8s: component: elasticsearch`` defines the pods that are part of the *elasticsearch* service in Kubernetes.
```
Comment (aanm) [inline Documentation/gettingstarted/elasticsearch.rst]: instead of

>``k8s: component: elasticsearch``

should be

> ``'k8s:component': elasticsearch``

**Review by aanm — DISMISSED** [review]

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
test-docs-please

**Pull request merged**